lints/lint_rsa_signature_algo_parameter_not_null.go

```diff
@@ -0,0 +1,118 @@
+
+func (l *rsaEncryptionParamNotNULL) Initialize() error {
+	return nil
+}
+
+func (l *rsaEncryptionParamNotNULL) CheckApplies(c *x509.Certificate) bool {
+	// explicitly check for util.OidRSAEncryption, as RSA-PSS or RSA-OAEP certificates might be classified with c.PublicKeyAlgorithm = RSA
+	return c.PublicKeyAlgorithmOID.Equal(util.OidRSAEncryption)
+}
+
+func (l *rsaEncryptionParamNotNULL) Execute(c *x509.Certificate) *LintResult {
+	// Try byte comparison for Algorithm
+	// Determine offset (SEQUENCE + length) first. Notably 1024 encodes length in 2 bytes, 2048 and up in 3 bytes
+	offset := getOffset(c.RawSubjectPublicKeyInfo)
+	if len(c.RawSubjectPublicKeyInfo) <= len(expectedSPKIAlgoBytes)+offset {
+		return &LintResult{Status: Fatal, Details: "error malformed pkixPublicKey"}
+	}
+
+	if bytes.Compare(c.RawSubjectPublicKeyInfo[offset:len(expectedSPKIAlgoBytes)+offset], expectedSPKIAlgoBytes) == 0 {
+		return &LintResult{Status: Pass}
+	}
+
+	input := cryptobyte.String(c.RawSubjectPublicKeyInfo)
+
+	var publicKeyInfo cryptobyte.String
+	if !input.ReadASN1(&publicKeyInfo, asn1_cryptobyte.SEQUENCE) {
+		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey"}
+	}
+
+	var algorithm cryptobyte.String
+	if !publicKeyInfo.ReadASN1(&algorithm, asn1_cryptobyte.SEQUENCE) {
+		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey algorithm"}
+	}
+
+	if !algorithm.SkipASN1(asn1_cryptobyte.OBJECT_IDENTIFIER) {
+		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey algorithm OID"}
+	}
+
+	if algorithm.Empty() {
+		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier missing required NULL parameter"}
+	}
+
+	var nullValue cryptobyte.String
+	if !algorithm.ReadASN1(&nullValue, asn1_cryptobyte.NULL) {
+		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier with non-NULL parameter"}
+	}
+
+	if len(nullValue) != 0 {
+		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier with NULL parameter containing trailing data"}
+	}
+
+	// ensure algorithm is empty and no trailing data is present
+	if !algorithm.Empty() {
+		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier with trailing data"}
+	}
+
+	return &LintResult{Status: Fatal, Details: "certificate rsa algorithm identifier appears correct, but didn't match byte-wise comparison"}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "e_rsa_encryption_parameter_not_null",
+		Description:   "RSA: Encoded algorithm identifier MUST have NULL parameters",
+		Citation:      "RFC 4055, Section 1.2",
+		Source:        RFC5280, // RFC4055 is referenced in RFC5280, Section 1
+		EffectiveDate: util.RFC5280Date,
+		Lint:          &rsaEncryptionParamNotNULL{},
+	})
+}
+
+func getOffset(asn1Sequence []byte) int {
+	if len(asn1Sequence) < 2 {
```

Comment:
Can you explain this clause? At face-value it seems like a case I'd expect to see an `error` value returned. Shouldn't the `rawSubjectPublicKeyInfo` `SEQUENCE` always be at least 1 tag byte and 1 length byte?

Should this lint also be enforcing the correct `rawSubjectPublicKeyInfo` `SEQUENCE` tag is present? I think the current implementation of `getOffset` will allow an invalid outer `SEQUENCE` as long as the DER is >= 2 bytes in length.

If we're less concerned about the outer `SEQUENCE` in this lint then maybe it would be acceptable to use `cryptobyte` to consume the `SEQUENCE` instead of using this `getOffset` function so that the lint doesn't need to concern itself with the length octet encoding and can just perform the bytewise comparison of the expected bytes? 

@sleevi What do you think?

Reply:
I think of this lint as being about "the encoding within the `AlgorithmIdentifier`", and thus less important about the context the `AlgorithmIdentifier` appears in.

That is, we can have AlgId in the `TBS.subjectPublicKeyInfo` and in the `TBS.signature` fields, and we should lint both of those. The outer certificates' `Certificate.signatureAlgorithm` is mutable by adversaries (since it's unsigned), so I'm less interested in that.

Under that framing, I think ignoring the outer `SEQUENCE` of the `subjectPublicKeyInfo` is fine, and I agree, using cryptobyte is probably a great (and safe) way to do that. If we're worried about trailing data on SPKI (and we should be, to be fair), I think that's probably as a separate lint, as we would trailing data in any other field?

Reply:
> That is, we can have AlgId in the TBS.subjectPublicKeyInfo and in the TBS.signature fields, and we should lint both of those.

@sleevi That makes sense. I think linting the `TBS.signature` field in this way will be mean needing to spelunk the coarse `RawTBSCertificate []byte` struct field with `cryptobyte` to find the alg id offset since there isn't a more targeted equiv. to `RawSubjectPublicKeyInfo` on the certificate struct given to the linter. I'm slightly in favour of keeping momentum on the existing `TBS.subjectPublicKeyInfo` focused lint and leaving a TODO for the Signature field because of that requirement for more new code. I could probably be persuaded its worth blocking on if you think its important to cover this field immediately.

@tadukurow What do you think? Do you have a pref. with regards to linting the signature RSA algorithm identifier in this branch vs in a follow-up? 

> Under that framing, I think ignoring the outer SEQUENCE of the subjectPublicKeyInfo is fine, and I agree, using cryptobyte is probably a great (and safe) way to do that.

Great :+1: Thanks for the input! I'll update my review feedback to encourage switching the `getOffset` code to equiv usage of `cryptobyte`.

>  If we're worried about trailing data on SPKI (and we should be, to be fair), I think that's probably as a separate lint, as we would trailing data in any other field?

Agreed, I think that would be a candidate for a separate lint.

Thanks again!

Reply:
So to clarify, we don't care about trailing bytes in the SPKI, including trailing bytes after the Parameter?

Reply:
> So to clarify, we don't care about trailing bytes in the SPKI, including trailing bytes after the Parameter?

That sounds reasonable to me :+1: A separate lint focused on trailing data could be added afterwards covering more fields.

Reply:
I think it’s pretty easy to check for trailing bytes here for the inner `SEQUENCE`, but the outer `SEQUENCE` is about an `RSAPublicKey` and that’s not what you’re linting here, if that makes sense.

So I’m a fan of checking after params but not algId

Reply:
Thought as much. I'll get that in and have a look at adding the lint for signatureAlgorithm.
Do we care much about the detailed error messages or is simple failure message enough?
Using CB to get offset and checking for trailing data, means the SPKI needs to be reparsed to get the detailed error messages. Removing that path would make it a lot simpler.

Reply:
I think the detailed error messages are valuable and the complexity isn't too great.